efi/preinstall/check_tpm.go

```diff
@@ -30,17 +31,190 @@ const (
+		// GetCapabilityTPMProperty, used above) which simplifies things here.
+		switch props, err := tpm.GetCapabilityPCRProperties(data.prop, 1); {
+		case err != nil:
+			return false, fmt.Errorf("cannot obtain value of %v: %w", data.prop, err)
+		case (len(props) < 1 || props[0].Tag != data.prop) && data.required:
+			return false, fmt.Errorf("cannot obtain value of %v: missing property", data.prop)
+		case len(props) < 1 || props[0].Tag != data.prop:
+			// ok
+		default:
+			// Filter out PCRs not associated with the S-RTM from the result.
+			var pcrs tpm2.PCRSelect
+			for _, pcr := range props[0].Select {
+				if pcr > 15 {
+					continue
+				}
+				pcrs = append(pcrs, pcr)
+			}
+
+			expectedBitmap, err := data.expected.ToBitmap(0)
+			if err != nil {
+				// The PCR selection is hard-coded.
+				panic(err)
+			}
+			bitmap, err := pcrs.ToBitmap(0)
+			if err != nil {
+				// This value was unmarshaled from a PCRSelectBitmap by
+				// go-tpm2 and we've also pruned the PCRs so there's no
+				// way this can fail.
+				panic(err)
+			}
+			if !bytes.Equal(bitmap.Bytes, expectedBitmap.Bytes) {
+				return false, nil
+			}
+		}
+	}
+
+	// Check commands.
+	for _, command := range []tpm2.CommandCode{
+		tpm2.CommandSelfTest,
+		tpm2.CommandGetTestResult,
+		tpm2.CommandStartAuthSession,
+		tpm2.CommandCreate,
+		tpm2.CommandLoad,
+		tpm2.CommandLoadExternal,
+		tpm2.CommandReadPublic,
+		tpm2.CommandUnseal,
+		tpm2.CommandObjectChangeAuth,
+		tpm2.CommandImport,
+		tpm2.CommandHashSequenceStart,
+		tpm2.CommandSequenceUpdate,
+		tpm2.CommandEventSequenceComplete,
+		tpm2.CommandVerifySignature,
+		tpm2.CommandPCRRead,
+		tpm2.CommandPolicySigned,
+		tpm2.CommandPolicySecret,
+		tpm2.CommandPolicyOR,
+		tpm2.CommandPolicyPCR,
+		tpm2.CommandPolicyNV,
+		tpm2.CommandPolicyCommandCode,
+		tpm2.CommandPolicyAuthorize,
+		tpm2.CommandPolicyAuthValue,
+		tpm2.CommandPolicyGetDigest,
+		tpm2.CommandPolicyNvWritten,
+		tpm2.CommandCreatePrimary,
+		tpm2.CommandClear,
+		tpm2.CommandClearControl,
+		tpm2.CommandHierarchyChangeAuth,
+		tpm2.CommandDictionaryAttackLockReset,
+		tpm2.CommandDictionaryAttackParameters,
+		tpm2.CommandContextSave,
+		tpm2.CommandContextLoad,
+		tpm2.CommandFlushContext,
+		tpm2.CommandEvictControl,
+		tpm2.CommandGetCapability,
+		tpm2.CommandNVDefineSpace,
+		tpm2.CommandNVUndefineSpace,
+		tpm2.CommandNVReadPublic,
+		tpm2.CommandNVWrite,
+		tpm2.CommandNVIncrement,
+		tpm2.CommandNVWriteLock,
+		tpm2.CommandNVRead,
+	} {
+		if !tpm.IsCommandSupported(command) {
+			return false, nil
+		}
+	}
+
+	// XXX: We can't check NV storage size, but we don't check that in the normal case either.
+	return true, nil
+}
+
 // checkTPM2DeviceFlags are passed to openAndCheckTPM2Device
 type checkTPM2DeviceFlags int
 
 const (
-	// checkTPM2DeviceInVM indicates that the current environment is a
-	// virtual machine.
-	checkTPM2DeviceInVM checkTPM2DeviceFlags = 1 << iota
```

Comment:
this goes away because we don't have to special case this anymore?

Reply:
That's right